File: src/services/quote.service.ts
import { QuoteRepository } from "@/repositories/quote.repository";
import { ProductRepository } from "@/repositories/product.repository";
import { CustomerService } from "@/services/customer.service";
import { type Quote, type QuoteStatus, type QuoteItem, type QuoteRoom, quoteSchema } from "@/types/quotes/quote";
import { AppError } from "@/middleware/error-handler";
import { logger } from "@/utils/logger";
import { ObjectId } from "mongodb";
import { Timestamp } from "firebase-admin/firestore";
import type { Currency } from "@/types/currency";

export class QuoteService {
  private repository: QuoteRepository;
  private productRepository: ProductRepository;
  private customerService: CustomerService;

  constructor() {
    this.repository = new QuoteRepository();
    this.productRepository = new ProductRepository();
    this.customerService = new CustomerService();
  }

  async createQuote(companyId: string, creatorId: string, creatorName: string, currency: Currency): Promise<Quote> {
    const quoteNumber = await this.repository.getNextQuoteNumber(companyId);

    const quote: Quote = quoteSchema.parse({
      companyId,
      quoteNumber,
      creatorId,
      creatorName,
      currency,
      rooms: [],
      status: "draft",
    });

    return await this.repository.create(companyId, quote);
  }

  async getQuote(companyId: string, id: string): Promise<Quote> {
    const quote = await this.repository.findById(companyId, id);
    if (!quote) {
      throw new AppError(404, "Quote not found", "QUOTE_NOT_FOUND");
    }
    return quote;
  }

  async updateQuoteCustomer(
    companyId: string,
    id: string,
    input: { customerId?: string; newCustomer?: { name: string; surname: string; phoneNumber?: string } }
  ): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    let customerId: string | undefined = input.customerId;
    let customerName: string | undefined;

    if (input.newCustomer) {
      const created = await this.customerService.createCustomer(companyId, {
        name: input.newCustomer.name,
        surname: input.newCustomer.surname,
        phoneNumber: input.newCustomer.phoneNumber,
        status: "active"
      } as any);
      customerId = created._id?.toString();
      customerName = `${created.name} ${created.surname}`;
    } else if (customerId) {
      const customer = await this.customerService.getCustomer(companyId, customerId);
      if (!customer) {
        throw new AppError(400, "Customer not found", "CUSTOMER_NOT_FOUND");
      }
      customerName = `${customer.name} ${customer.surname}`;
    }

    const updated = await this.repository.update(companyId, id, {
      customerId: customerId ? new ObjectId(customerId) : undefined,
      customerName,
    });

    return updated!;
  }

  async updateQuoteCurrency(companyId: string, id: string, currency: Currency): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    // If currency changes, we need to reset conversions or keep only the new base as 1
    const conversions = { [currency]: 1 };

    const updated = await this.repository.update(companyId, id, {
      currency,
      conversions: conversions as any,
    });

    return await this.recalculateQuoteTotal(companyId, id);
  }

  async updateQuoteConversions(companyId: string, id: string, conversions: Record<string, number>): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    // Ensure base currency rate is 1
    conversions[quote.currency] = 1;

    await this.repository.update(companyId, id, { conversions: conversions as any });

    return await this.recalculateQuoteTotal(companyId, id);
  }

  async addRoom(companyId: string, id: string, name: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    const room: QuoteRoom = {
      id: new ObjectId().toHexString(),
      name,
      items: [],
      total: 0,
    };

    const updated = await this.repository.update(companyId, id, {
      rooms: [...quote.rooms, room],
    });

    return updated!;
  }

  async deleteRoom(companyId: string, id: string, roomId: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    const roomExists = quote.rooms.some(r => r.id === roomId);
    if (!roomExists) {
      throw new AppError(404, "Room not found", "ROOM_NOT_FOUND");
    }

    const updatedRooms = quote.rooms.filter(r => r.id !== roomId);

    await this.repository.update(companyId, id, { rooms: updatedRooms });
    return await this.recalculateQuoteTotal(companyId, id);
  }

  async updateRoomName(companyId: string, id: string, roomId: string, name: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    const room = quote.rooms.find(r => r.id === roomId);
    if (!room) {
      throw new AppError(404, "Room not found", "ROOM_NOT_FOUND");
    }

    room.name = name;

    const updated = await this.repository.update(companyId, id, { rooms: quote.rooms });
    return updated!;
  }

  async addItemsToRoom(
    companyId: string,
    id: string,
    roomId: string,
    itemsInput: { productId: string; quantity: number; customPrice?: number }[]
  ): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    if (!quote.conversions || Object.keys(quote.conversions).length <= 1 && !quote.conversions[quote.currency]) {
      // Technically we initialized it with base=1, but let's check if others are needed
    }

    const room = quote.rooms.find(r => r.id === roomId);
    if (!room) {
      throw new AppError(404, "Room not found", "ROOM_NOT_FOUND");
    }

    const newItems: QuoteItem[] = [];

    for (const input of itemsInput) {
      const product = await this.productRepository.findById(input.productId);
      if (!product) {
        throw new AppError(400, `Product not found: ${input.productId}`, "PRODUCT_NOT_FOUND");
      }

      const conversionRate = quote.conversions[product.currency];
      if (!conversionRate) {
        throw new AppError(400, `Conversion rate for ${product.currency} is not defined in this quote.`, "CONVERSION_MISSING");
      }

      const unitPrice = input.customPrice ?? product.price;
      const convertedUnitPrice = unitPrice * conversionRate;
      const totalPrice = input.quantity * convertedUnitPrice;

      newItems.push({
        id: new ObjectId().toHexString(),
        productId: product._id!,
        name: product.name,
        quantity: input.quantity,
        unitPrice,
        originalCurrency: product.currency,
        convertedUnitPrice,
        totalPrice,
      });
    }

    room.items.push(...newItems);

    await this.repository.update(companyId, id, { rooms: quote.rooms });
    return await this.recalculateQuoteTotal(companyId, id);
  }

  async updateItem(
    companyId: string,
    id: string,
    roomId: string,
    itemId: string,
    updates: { quantity?: number; customPrice?: number }
  ): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    const room = quote.rooms.find(r => r.id === roomId);
    if (!room) throw new AppError(404, "Room not found", "ROOM_NOT_FOUND");

    const item = room.items.find(i => i.id === itemId);
    if (!item) throw new AppError(404, "Item not found", "ITEM_NOT_FOUND");

    if (updates.quantity !== undefined) item.quantity = updates.quantity;
    if (updates.customPrice !== undefined) item.unitPrice = updates.customPrice;

    // Recalculate item prices
    const conversionRate = quote.conversions[item.originalCurrency];
    item.convertedUnitPrice = item.unitPrice * conversionRate;
    item.totalPrice = item.quantity * item.convertedUnitPrice;

    await this.repository.update(companyId, id, { rooms: quote.rooms });
    return await this.recalculateQuoteTotal(companyId, id);
  }

  async removeItem(companyId: string, id: string, roomId: string, itemId: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    const room = quote.rooms.find(r => r.id === roomId);
    if (!room) throw new AppError(404, "Room not found", "ROOM_NOT_FOUND");

    room.items = room.items.filter(i => i.id !== itemId);

    await this.repository.update(companyId, id, { rooms: quote.rooms });
    return await this.recalculateQuoteTotal(companyId, id);
  }

  async submitForApproval(companyId: string, id: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    this.ensureEditable(quote);

    // Validation
    if (!quote.customerId) {
      throw new AppError(400, "Customer must be selected before submitting for approval.", "VALIDATION_ERROR");
    }

    if (quote.rooms.length === 0 || !quote.rooms.some(r => r.items.length > 0)) {
      throw new AppError(400, "Quote must have at least one room with at least one product.", "VALIDATION_ERROR");
    }

    if (quote.total <= 0) {
      throw new AppError(400, "Quote total must be greater than 0.", "VALIDATION_ERROR");
    }

    const updated = await this.repository.update(companyId, id, { status: "sent_for_approval" });
    return updated!;
  }

  async approveQuote(companyId: string, id: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    if (quote.status !== "sent_for_approval") {
      throw new AppError(400, "Only quotes sent for approval can be approved.", "INVALID_STATUS");
    }

    const updated = await this.repository.update(companyId, id, { status: "approved" });
    return updated!;
  }

  async denyQuote(companyId: string, id: string): Promise<Quote> {
    const quote = await this.getQuote(companyId, id);
    if (quote.status !== "sent_for_approval") {
      throw new AppError(400, "Only quotes sent for approval can be denied.", "INVALID_STATUS");
    }

    const updated = await this.repository.update(companyId, id, { status: "denied" });
    return updated!;
  }

  async listQuotes(companyId: string, userId: string, role: string): Promise<Quote[]> {
    const isAdmin = role === "admin" || role === "sudo";
    return await this.repository.findAll(companyId, isAdmin ? {} : { creatorId: userId });
  }

  private async recalculateQuoteTotal(companyId: string, id: string): Promise<Quote> {
    const quote = await this.repository.findById(companyId, id);
    if (!quote) return null as any;

    let grandTotal = 0;

    for (const room of quote.rooms) {
      let roomTotal = 0;
      for (const item of room.items) {
        // Re-apply conversion in case rates changed
        const rate = quote.conversions[item.originalCurrency] || 1;
        item.convertedUnitPrice = item.unitPrice * rate;
        item.totalPrice = item.quantity * item.convertedUnitPrice;
        roomTotal += item.totalPrice;
      }
      room.total = roomTotal;
      grandTotal += roomTotal;
    }

    const updated = await this.repository.update(companyId, id, {
      rooms: quote.rooms,
      total: grandTotal,
    });

    return updated!;
  }

  private ensureEditable(quote: Quote) {
    if (quote.status === "approved" || quote.status === "denied") {
      throw new AppError(400, "Approved or Denied quotes cannot be edited.", "READ_ONLY_ERROR");
    }
  }
}

File: fisoft/src/main/kotlin/services/QuoteService.kt
package com.ilyasdemirkiran.services

import com.ilyasdemirkiran.middleware.AppError
import com.ilyasdemirkiran.repository.ProductRepository
import com.ilyasdemirkiran.repository.QuoteRepository
import com.ilyasdemirkiran.types.Money
import com.ilyasdemirkiran.types.quotes.*
import io.ktor.http.*
import java.time.Instant
import java.util.UUID

class QuoteService(
    private val repo: QuoteRepository = QuoteRepository(),
    private val productRepo: ProductRepository = ProductRepository(),
    private val customerService: CustomerService = CustomerService()
) {
    private fun ensureEditable(quote: Quote) {
        if (quote.status == "approved" || quote.status == "denied") {
            throw AppError(HttpStatusCode.BadRequest, "Approved or Denied quotes cannot be edited.", "READ_ONLY_ERROR")
        }
    }

    private suspend fun recalculateAndSave(companyId: String, quote: Quote): Quote {
        val updatedRooms = quote.rooms.map { room ->
            val roomTotal = room.items.fold(0.0) { acc, item -> acc + (item.totalPrice ?: 0.0) }
            room.copy(total = roomTotal)
        }

        val grandTotalDouble = updatedRooms.fold(0.0) { acc, r -> acc + (r.total ?: 0.0) }
        val grandTotal = Money.fromDouble(grandTotalDouble)
        val totalAfterDiscount = grandTotal.applyDiscountPercent(quote.discountPercent ?: 0.0)

        val updated = quote.copy(
            rooms = updatedRooms,
            total = grandTotal,
            totalAfterDiscount = totalAfterDiscount
        )

        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun createQuote(companyId: String, creatorId: String, creatorName: String, currency: String = "TRY", conversions: QuoteConversions = QuoteConversions()): Quote {
        val quoteNumber = repo.getNextQuoteNumber(companyId)
        val quote = Quote(
            companyId = companyId,
            quoteNumber = quoteNumber,
            creatorId = creatorId,
            creatorName = creatorName,
            currency = currency,
            conversions = conversions,
            rooms = emptyList(),
            status = "draft",
            createdAt = Instant.now().toString()
        )
        return repo.create(companyId, quote)
    }

    suspend fun getQuote(companyId: String, id: String): Quote {
        return repo.findById(companyId, id)
            ?: throw AppError(HttpStatusCode.NotFound, "Quote not found", "QUOTE_NOT_FOUND")
    }

    suspend fun updateQuoteCustomer(companyId: String, id: String, customerId: String?, customerName: String?): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)
        val updated = quote.copy(customerId = customerId, customerName = customerName)
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun updateQuoteCurrency(companyId: String, id: String, currency: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)
        val updated = quote.copy(currency = currency)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun updateQuoteConversions(companyId: String, id: String, conversions: QuoteConversions): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)
        val updated = quote.copy(conversions = conversions)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun updateDiscountPercent(companyId: String, id: String, discountPercent: Double): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)
        val updated = quote.copy(discountPercent = discountPercent)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun addRoom(companyId: String, id: String, name: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val room = QuoteRoom(
            id = UUID.randomUUID().toString(),
            name = name,
            items = emptyList(),
            total = 0.0
        )

        val updated = quote.copy(rooms = quote.rooms + room)
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun updateRoomName(companyId: String, id: String, roomId: String, name: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) r.copy(name = name) else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun deleteRoom(companyId: String, id: String, roomId: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val updatedRooms = quote.rooms.filter { it.id != roomId }
        val updated = quote.copy(rooms = updatedRooms)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun addItemsToRoom(companyId: String, id: String, roomId: String, itemsInput: List<Pair<String, Double>>): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val room = quote.rooms.find { it.id == roomId }
            ?: throw AppError(HttpStatusCode.NotFound, "Room not found", "ROOM_NOT_FOUND")

        val newItems = mutableListOf<QuoteItem>()
        for ((productId, quantity) in itemsInput) {
            val product = productRepo.findById(productId)
                ?: throw AppError(HttpStatusCode.BadRequest, "Product not found: $productId", "PRODUCT_NOT_FOUND")

            val conversionRate = when (product.currency) {
                "USD" -> quote.conversions?.USD ?: 1.0
                "EUR" -> quote.conversions?.EUR ?: 1.0
                else -> quote.conversions?.TRY ?: 1.0
            }

            val unitPrice = product.price.toDouble
            val convertedUnitPrice = unitPrice * conversionRate
            val totalPrice = convertedUnitPrice * quantity

            newItems.add(
                QuoteItem(
                    id = UUID.randomUUID().toString(),
                    productId = product._id,
                    name = product.name,
                    publicName = "",
                    quantity = quantity,
                    unitPrice = unitPrice,
                    originalCurrency = product.currency,
                    convertedUnitPrice = convertedUnitPrice,
                    totalPrice = totalPrice
                )
            )
        }

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) r.copy(items = r.items + newItems) else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun addCustomItemToRoom(
        companyId: String,
        id: String,
        roomId: String,
        name: String,
        quantity: Double,
        unitPrice: Double,
        currency: String
    ): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val conversionRate = when (currency) {
            "USD" -> quote.conversions?.USD ?: 1.0
            "EUR" -> quote.conversions?.EUR ?: 1.0
            else -> quote.conversions?.TRY ?: 1.0
        }

        val convertedUnitPrice = unitPrice * conversionRate
        val totalPrice = convertedUnitPrice * quantity

        val newItem = QuoteItem(
            id = UUID.randomUUID().toString(),
            name = name,
            publicName = "",
            quantity = quantity,
            unitPrice = unitPrice,
            originalCurrency = currency,
            convertedUnitPrice = convertedUnitPrice,
            totalPrice = totalPrice
        )

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) r.copy(items = r.items + newItem) else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun updateItem(
        companyId: String,
        id: String,
        roomId: String,
        itemId: String,
        quantity: Double?,
        customPrice: Double?
    ): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) {
                val updatedItems = r.items.map { item ->
                    if (item.id == itemId) {
                        val newQty = quantity ?: (item.quantity ?: 0.0)
                        val newPrice = customPrice ?: (item.unitPrice ?: 0.0)

                        val conversionRate = when (item.originalCurrency) {
                            "USD" -> quote.conversions?.USD ?: 1.0
                            "EUR" -> quote.conversions?.EUR ?: 1.0
                            else -> quote.conversions?.TRY ?: 1.0
                        }

                        val convertedUnitPrice = newPrice * conversionRate
                        val totalPrice = newQty * convertedUnitPrice

                        item.copy(
                            quantity = newQty,
                            unitPrice = newPrice,
                            convertedUnitPrice = convertedUnitPrice,
                            totalPrice = totalPrice
                        )
                    } else item
                }
                r.copy(items = updatedItems)
            } else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun updateItemPublicName(companyId: String, id: String, roomId: String, itemId: String, publicName: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) {
                val updatedItems = r.items.map { item ->
                    if (item.id == itemId) item.copy(publicName = publicName) else item
                }
                r.copy(items = updatedItems)
            } else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun removeItem(companyId: String, id: String, roomId: String, itemId: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        val updatedRooms = quote.rooms.map { r ->
            if (r.id == roomId) {
                r.copy(items = r.items.filter { it.id != itemId })
            } else r
        }

        val updated = quote.copy(rooms = updatedRooms)
        return recalculateAndSave(companyId, updated)
    }

    suspend fun submitForApproval(companyId: String, id: String): Quote {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)

        if (quote.customerId == null) {
            throw AppError(HttpStatusCode.BadRequest, "Customer must be selected before submitting for approval.", "VALIDATION_ERROR")
        }
        if (quote.rooms.isEmpty() || !quote.rooms.any { it.items.isNotEmpty() }) {
            throw AppError(HttpStatusCode.BadRequest, "Quote must have at least one room with at least one product.", "VALIDATION_ERROR")
        }

        val updated = quote.copy(status = "sent_for_approval")
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun approveQuote(companyId: String, id: String): Quote {
        val quote = getQuote(companyId, id)
        if (quote.status != "sent_for_approval") {
            throw AppError(HttpStatusCode.BadRequest, "Only quotes sent for approval can be approved.", "INVALID_STATUS")
        }
        val updated = quote.copy(status = "approved")
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun denyQuote(companyId: String, id: String): Quote {
        val quote = getQuote(companyId, id)
        if (quote.status != "sent_for_approval") {
            throw AppError(HttpStatusCode.BadRequest, "Only quotes sent for approval can be denied.", "INVALID_STATUS")
        }
        val updated = quote.copy(status = "denied")
        return repo.updateQuote(companyId, updated) ?: updated
    }

    suspend fun listQuotes(companyId: String, userId: String, isAdmin: Boolean): List<Quote> {
        return repo.findAll(companyId, if (isAdmin) null else userId)
    }

    suspend fun deleteQuote(companyId: String, id: String) {
        val quote = getQuote(companyId, id)
        ensureEditable(quote)
        repo.delete(companyId, id)
    }
}
